README.md

# iNZightTS2: Time Series Visualisation and Forecasting

[![R-CMD-check](https://github.com/iNZightVIT/iNZightTS2/workflows/R-CMD-check/badge.svg)](https://github.com/iNZightVIT/iNZightTS2/actions?workflow=R-CMD-check)
[![Codecov test
coverage](https://codecov.io/gh/iNZightVIT/iNZightTS2/branch/dev/graph/badge.svg)](https://codecov.io/gh/iNZightVIT/iNZightTS2?branch=dev)
[![License: GPL
v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)
[![CRAN](https://www.r-pkg.org/badges/version/iNZightTS)](https://CRAN.R-project.org/package=iNZightTS)

<!-- [![CRAN](https://www.r-pkg.org/badges/version/iNZightTS2)](https://CRAN.R-project.org/package=iNZightTS2) -->

## Overview

The iNZightTS2 package provides some simple analysis tools for exploring
time series data, which is used by
[iNZight](https://github.com/iNZightVIT/iNZight). The package uses
[tidyverts](https://tidyverts.org/) to store and process time series
data and [ggplot2](https://github.com/tidyverse/ggplot2) to produce
customisable graphics.

## Installation

<!-- The _stable_ version can be installed from [CRAN](https://cran.r-project.org/web/packages/iNZightTS2/index.html):


```r
install.packages("iNZightTS2")
```
-->

Install from [GitHub](https://github.com/iNZightVIT/iNZightTS2):

``` r
remotes::install_github("iNZightVIT/iNZightTS2")
```

## Usage

``` r
library(iNZightTS2)
library(dplyr)
library(tidyr)
```

### Getting started

Use the `inzightts` function to create a temporal data set to be used by
the functions of the package. The data is stored as a
[tsibble](https://github.com/tidyverts/tsibble).

``` r
data <- visitorsQ |>
    pivot_longer(!Date, names_to = "Country", values_to = "Visitors") |>
    inzightts(key = "Country")
data
#> # A tsibble: 216 x 3 [1Q]
#> # Key:       Country [4]
#>      index Country   Visitors
#>      <qtr> <chr>        <int>
#>  1 1998 Q4 Australia    20288
#>  2 1999 Q1 Australia    22047
#>  3 1999 Q2 Australia    14362
#>  4 1999 Q3 Australia    15775
#>  5 1999 Q4 Australia    21209
#>  6 2000 Q1 Australia    25261
#>  7 2000 Q2 Australia    15891
#>  8 2000 Q3 Australia    17117
#>  9 2000 Q4 Australia    22761
#> 10 2001 Q1 Australia    27539
#> # … with 206 more rows
```

### Graphics

Exploratory analysis of time series data starts from a smoothed line
plot.

``` r
plot(data)
```

<img src="man/figures/raw-plot-1.png" width="100%" />

### Decompositions

Time series data often exhibit features such as trend, season and cycle.
A decomposition plot breaks the data into visual components which
simplies the analysis.

``` r
dcmp <- data |>
    filter(Country == "Australia") |>
    inzightts() |>
    decomp()
dcmp
#> # A dable: 54 x 8 [1Q]
#> # Key:     Country, .model [1]
#> # :        Visitors = trend + season_year + remainder
#>    Country   .model                                     index Visit…¹  trend seaso…² remai…³ seaso…⁴
#>  * <chr>     <chr>                                      <qtr>   <dbl>  <dbl>   <dbl>   <dbl>   <dbl>
#>  1 Australia feasts::STL(Visitors ~ trend() + season… 1998 Q4   20288 16858.   2191.   1239.  18097.
#>  2 Australia feasts::STL(Visitors ~ trend() + season… 1999 Q1   22047 17586.   7515.  -3054.  14532.
#>  3 Australia feasts::STL(Visitors ~ trend() + season… 1999 Q2   14362 18173.  -5731.   1920.  20093.
#>  4 Australia feasts::STL(Visitors ~ trend() + season… 1999 Q3   15775 18928.  -3975.    822.  19750.
#>  5 Australia feasts::STL(Visitors ~ trend() + season… 1999 Q4   21209 19248.   2191.   -230.  19018.
#>  6 Australia feasts::STL(Visitors ~ trend() + season… 2000 Q1   25261 19601.   7515.  -1855.  17746.
#>  7 Australia feasts::STL(Visitors ~ trend() + season… 2000 Q2   15891 20127.  -5731.   1496.  21622.
#>  8 Australia feasts::STL(Visitors ~ trend() + season… 2000 Q3   17117 20652.  -3975.    439.  21092.
#>  9 Australia feasts::STL(Visitors ~ trend() + season… 2000 Q4   22761 20972.   2191.   -402.  20570.
#> 10 Australia feasts::STL(Visitors ~ trend() + season… 2001 Q1   27539 21528.   7515.  -1504.  20024.
#> # … with 44 more rows, and abbreviated variable names ¹​Visitors, ²​season_year, ³​remainder,
#> #   ⁴​season_adjust
```

``` r
plot(dcmp, title = "Visitors to Australia")
```

<img src="man/figures/decomp-plot-1.png" width="100%" />

### Visualising seasonal effects

Plots are helpful in revealing the underlying seasonal pattern of time
series data.

``` r
seasonplot(data)
```

<img src="man/figures/season-plot-1.png" width="100%" />

``` r
subseries(data)
```

<img src="man/figures/subseries-plot-1.png" width="100%" />

### Forecasting

``` r
pred <- predict(data)
summary(pred)
#> 
#> 95% Prediction Interval
#> # A tsibble: 32 x 5 [1Q]
#> # Key:       Country [4]
#>    Country                        Time Fitted  Lower  Upper
#>    <chr>                         <qtr>  <dbl>  <dbl>  <dbl>
#>  1 Australia                   2012 Q2 25550. 23052. 28047.
#>  2 Australia                   2012 Q3 30625. 27790. 33461.
#>  3 Australia                   2012 Q4 38674. 35748. 41600.
#>  4 Australia                   2013 Q1 43773. 40821. 46724.
#>  5 Australia                   2013 Q2 26915. 23421. 30409.
#>  6 Australia                   2013 Q3 31797. 28162. 35433.
#>  7 Australia                   2013 Q4 39874. 36199. 43549.
#>  8 Australia                   2014 Q1 44769. 41082. 48456.
#>  9 China..People.s.Republic.of 2012 Q2  8069.  6999.  9138.
#> 10 China..People.s.Republic.of 2012 Q3  8624.  6704. 10544.
#> 11 China..People.s.Republic.of 2012 Q4  9797.  7199. 12395.
#> 12 China..People.s.Republic.of 2013 Q1 11054.  7964. 14145.
#> 13 China..People.s.Republic.of 2013 Q2  8710.  4944. 12477.
#> 14 China..People.s.Republic.of 2013 Q3  8957.  4581. 13333.
#> 15 China..People.s.Republic.of 2013 Q4  9874.  5038. 14710.
#> 16 China..People.s.Republic.of 2014 Q1 11139.  5991. 16287.
#> 17 Japan                       2012 Q2  2106.  1267.  2945.
#> 18 Japan                       2012 Q3  3239.  2052.  4425.
#> 19 Japan                       2012 Q4  2789.  1336.  4242.
#> 20 Japan                       2013 Q1  4017.  2340.  5695.
#> 21 Japan                       2013 Q2  1719.  -397.  3835.
#> 22 Japan                       2013 Q3  2851.   374.  5329.
#> 23 Japan                       2013 Q4  2401.  -391.  5194.
#> 24 Japan                       2014 Q1  3630.   554.  6706.
#> 25 United.Kingdom              2012 Q2  9296.  6244. 12349.
#> 26 United.Kingdom              2012 Q3  8283.  4829. 11737.
#> 27 United.Kingdom              2012 Q4 17986. 14450. 21522.
#> 28 United.Kingdom              2013 Q1 28182. 24566. 31798.
#> 29 United.Kingdom              2013 Q2  7784.  2531. 13037.
#> 30 United.Kingdom              2013 Q3  6771.  1006. 12535.
#> 31 United.Kingdom              2013 Q4 16474. 10514. 22434.
#> 32 United.Kingdom              2014 Q1 26670. 20520. 32820.
#> 
#> Model:
#> # A mable: 4 x 2
#> # Key:     Country [4]
#>   Country                                              Visitors
#>   <chr>                                                 <model>
#> 1 Australia                   <ARIMA(1,0,0)(1,1,0)[4] w/ drift>
#> 2 China..People.s.Republic.of          <ARIMA(2,0,0)(1,1,0)[4]>
#> 3 Japan                                <ARIMA(0,1,0)(0,1,1)[4]>
#> 4 United.Kingdom                       <ARIMA(0,1,2)(0,1,0)[4]>
```

``` r
plot(pred)
```

<img src="man/figures/forecast-plot-1.png" width="100%" />


NEWS.Md
# iNZightTS2 2.0.0

This is a new package based off the original [iNZightTS](github.com/iNZightVIT/iNZightTS) package. Versioning will continue, as the long-term plan is to merge these changes into the original package once it is fully deprecated.

- rebuild package with features from **tidyverts**
- new: sub-series plot
- enhanced: introduce key for multiple time series
- enhanced: changed the default behaviour of multiple variables in time series plots
- added readme with example usage (#56)

---

**Previous versions of the original `iNZightTS` package are listed here.**

# iNZightTS 1.5.9

- fix bug that occurs if data passed to `iNZightTS()` is a tibble

# iNZightTS 1.5.8

- change arguments to avoid deprecation arguments with next ggplot2 release (#62)

# iNZightTS 1.5.7

- print comparison plots so they are always shown in the graphics device

# iNZightTS 1.5.6

- add colour legend to subtitle of decomposition plot

# iNZightTS 1.5.5

- fix bug preventing basic plots from working for annual data

# iNZightTS 1.5.4

- fix bug which meant passing `recompose = TRUE` to the plot method for decomposition objects was being ignored

# iNZightTS 1.5.3

- fix small bug in detecting time information where `[` subset operator returns tibble/dataframe (not vector)

# iNZightTS 1.5.2

- disable multiplicative series if any values are zero or lower
- add appropriate subtitle to trend panel of decomposition plot
- fix a bug preventing `iNZightTS` from accepting a valid `stats::ts` object
- specify `stringsAsFactors = TRUE` for upcoming R 4.0.0
- fix documentation for CRAN

# iNZightTS 1.5.1

**Release date**: 14 November 2019

- fix a bug in `seasonplot()` where a strange interaction between Windows' `Cairo()` graphics device and `egg::ggarrange()` was resulting in the graphs not being displayed

# iNZightTS 1.5

**Release date**: 11 November 2019

- add argument to `plot()` functions to show smoother (default is `TRUE`)
- add UTF-8 encoding to DESCRIPTION file as requested by roxygen :)
- add ggplot2 version of season plot and decompose plot
- for compare plot with additive effects, make main and effects y-limits equally scaled
- fix bug where smoothing parameter wasn't being passed to compare effects graph

# iNZightTS 1.4.3

**Release date**: 2 September 2019

- automatically detect 5-day week data (as opposed to 7-day week)

# iNZightTS 1.4.2

**Release date**: 15 July 2019

- specify (time/horizontal) axes limits to view part of a time series (#16)
- use only part of a series to create model + forecast (#17)
- recode forecast plot using `ggplot2`

# iNZightTS 1.4.1

**Release date**: 29 May 2018

- add the seasonal lines for each cycle back into the separate-series comparison plot

# iNZightTS 1.4

**Release date**: 23 January 2018

- **ggplot2** is now being used for the (majority of) time series graphs
- various bugfixes

# iNZightTS 1.3-1

**Release date**: 20 October 2017

- Modify missing-values error message

# iNZightTS 1.3

**Release date**: 02 October 2017

- Bug fixes and package improvements to pass CRAN checks

# iNZightTS 1.2

**Release date**: 24 March 2017

- Main change is addition of smoothing parameter to plots
- Improved time variable detection, including weekly, daily, hourly data
- Better use of x- and y-axis labels
- Other minor bug fixes

# iNZightTS 1.1.7

**Release date**: 23 September 2016

- catch errors trying to guess TS structure

# iNZightTS 1.1.6

**Release date**: 02 February 2015

- reposition the axis label

# iNZightTS 1.1.5

**Release date**: 12 January 2015

- fixed a missed function export for `compareplot`

# iNZightTS 1.1.3

**Release date**: 17 September 2014

- added x label to additional plots

# iNZightTS 1.1.2

**Release date**: 4 April 2014

- y-axis label support now on all plots.
- Various bug fixes to accommodate multiplicative time series.

# iNZightTS 1.1.1

**Release date**: 27 March 2014

- New functionality using proportion values to calculate `x`,
  and calculated using order to control the x-axis.

- Changed some graphical parameter settings to avoid
  over-writing objects with a "white" background.

# iNZightTS 1.1

**Release date**: 18 January 2014

- No user-visible changes to the package, however re-released with the
  launch of iNZight 2.0.


forecastplot.R
##' Plot a raw time series together with it's fitted curve and add
##' forecasts and prediction intervals to the end.
##'
##' The predictions and prediction intervals are the result of models
##' fitted by the Holt-Winters method. The amount of predicted
##' observations is calculated by 2 * \code{freq}, where \code{freq} is
##' the frequency of the time series object.
##'
##' @title Forecast plot
##'
##' @param vars \code{iNZightTS} object
##' 
##' @param multiplicative logical. If \code{TRUE}, a multiplicative model is used,
##' otherwise an additive model is used by default.
##'
##' @param ylab a title for the y axis
##'
##' @param show logical. If \code{TRUE}, draw the plot by default.
##'
##' @return A multiple time series of the predicted values with columns fit,
##' lwr and upr for the predicted values and the lower and upper bounds
##' respectively.
##'
##' @references C.C Holt (1957)
##' Forecasting seasonals and trends by exponentially weighted
##' moving averages,
##' ONR Research Memorandum, Carnigie Institute 52.
##'
##' P.R Winters (1960)
##' Forecasting sales by exponentially weighted moving averages,
##' \emph{Management Science} \bold{6}, 324--342.
##'
##' @seealso \code{\link{iNZightTS}},
##' \code{\link{HoltWinters}}
##'
##' @examples z <- iNZightTS(ldeaths)
##' forecastplot(z)
##' 
##' @export
forecastplot <-
    function(vars, ylab="", multiplicative = FALSE, show = TRUE) {
      if (as.integer(vars$freq) == 1)
        return("forecastplot need a time series object with more than 1 frequency.")
    width <- 7
    height <- 4
    obs.col <- "black"
    obs.lwd <- 2
    fit.col <- "#0e8c07"
    pred.col <- "#b50000"
    int.col <- "#ff7777"
    band.col <- "#ffdbdb"
    tsObj <- vars$tsObj
    x.vals <- get.x(tsObj)
    ahead <- 2 * vars$freq
    seasonal.add = ifelse(multiplicative, "multiplicative",
                          "additive")
    AtsObj <- tsObj
    if (multiplicative)
        AtsObj <- log(tsObj)

      hw.fit <- try(HoltWinters(AtsObj), TRUE)
      if (inherits(hw.fit, "try-error")) stop("Holt-Winters could not converge.")
      
    temp.fit <- hw.fit
    pred <- predict(hw.fit, n.ahead = ahead, prediction.interval = TRUE)
    if (multiplicative){
    hw.fit$fitted <- exp(hw.fit$fitted)
    pred <- exp(pred)
    }
    x.pred.start <- tail(x.vals$x, 1)
    x.pred <- seq(from = x.pred.start, to = x.pred.start + ahead/vars$freq,
                  by = 1/vars$freq)
    y.pred <- c(tail(hw.fit$fitted[, "xhat"], 1), pred[, "fit"])
    ylims <- range(hw.fit$fitted[, "xhat"], pred, vars$tsObj)
    xlims <- c(x.vals$x[1], tail(x.pred, 1))
    xlims <- xlims + diff(xlims) * 0.02 * c(-1, 1)
    ylims <- ylims + diff(ylims) * c(-0.02, 0.15)
    x.fit <- get.x(hw.fit$fitted[, "xhat"])
    layout <- grid.layout(3, 3, widths = unit(c(1.5, 1, 1),
                                              c("inches", "null", "cm")), heights = unit(c(1.5, 1,
                                                                                           2), c("cm", "null", "cm")))
    parent.vp <- viewport(name = "parent", layout = layout)
    vps <- vpList()
    vps$plot <- viewport(name = "plot", layout.pos.row = 2,
                         layout.pos.col = 2, xscale = xlims, yscale = ylims,
                         default.units = "native")
    vps$left <- viewport(name = "left", layout.pos.row = 2,
                         layout.pos.col = 1)
    vps$right <- viewport(name = "right", layout.pos.row = 2,
                          layout.pos.col = 3)
    vps$top <- viewport(name = "top", layout.pos.row = 1, layout.pos.col = 1:3)
    vps$bottom <- viewport(name = "bottom", layout.pos.row = 3,
                           layout.pos.col = 1:3)
    vp.tree <- vpTree(parent.vp, vps)
    grobs <- gList()
    main <- paste("Holt-Winters", ifelse(multiplicative, "Multiplicative", "Additive"),
                  "prediction for", vars$currVar)
    grobs$title <- textGrob(main, gp = gpar(cex = 1.3, fontface = "bold"),
                            vp = vpPath("parent", "top"), name = "title")
    grobs$box <- rectGrob(vp = vpPath("parent", "plot"), name = "boxRect")
    grobs$sep <- linesGrob(tail(x.vals$x.units, 1), unit(ylims,
                                                         "native"), gp = gpar(col = "#555555", lty = "dashed"),
                           vp = vpPath("parent", "plot"), name = "separator")
    grobs$obsLine <- linesGrob(x.vals$x.units, unit(tsObj@.Data,
                                                    "native"), gp = gpar(col = obs.col, lwd = obs.lwd),
                               vp = vpPath("parent", "plot"), name = "obsLine")
    grobs$fitLine <- linesGrob(x.fit$x.units, unit(hw.fit$fitted[,
                                                                 "xhat"], "native"), gp = gpar(col = fit.col), name = "fitLine",
                               vp = vpPath("parent", "plot"))
    grobs$polygon <- polygonGrob(unit(c(x.pred[-1], rev(x.pred[-1])),
                                      "native"), unit(c(pred[, "lwr"], rev(pred[, "upr"])),
                                                      "native"), gp = gpar(fill = band.col, col = "transparent"),
                                 name = "polygon", vp = vpPath("parent", "plot"))
    grobs$predLine <- linesGrob(unit(x.pred, "native"), unit(y.pred,
                                                             "native"), gp = gpar(col = pred.col), name = "predLine",
                                vp = vpPath("parent", "plot"))
    grobs$upperLine <- linesGrob(unit(x.pred[-1], "native"),
                                 unit(pred[, "upr"], "native"), gp = gpar(col = int.col,
                                                                          lty = "dashed"), name = "upperPredLine", vp = vpPath("parent",
                                                                                                                               "plot"))
    grobs$lowerLine <- linesGrob(unit(x.pred[-1], "native"),
                                 unit(pred[, "lwr"], "native"), gp = gpar(col = int.col,
                                                                          lty = "dashed"), name = "lowerPredLine", vp = vpPath("parent",
                                                                                                                               "plot"))
    grobs$yAxis <- yaxisGrob(vp = vpPath("parent", "plot"),
                             name = "yAxis", gp = gpar(cex = 0.9))
    grobs$xAxis <- xaxisGrob(vp = vpPath("parent", "plot"),
                             name = "xAxis", gp = gpar(cex = 0.9))
    grobs$xLabel <- textGrob("Time", y = 0.3, name = "xLabel",
                             vp = vpPath("parent", "bottom"))
    grobs$yLabel <- textGrob(ylab, x = 0, y = 0.5, name = "yLabel",
                             rot = 90, vjust = -5.5, vp = vpPath("parent", "plot"))
    grobs$rawKey <- linesGrob(x = unit(c(2, 8), "mm"), y = unit(1,
                                                                "npc") - unit(0.6, "lines"), vp = vpPath("parent", "plot"),
                              name = "rawKey", gp = gpar(col = obs.col, lwd = 3))
    grobs$rawKeyText <- textGrob(label = "Raw data", x = unit(10,
                                                              "mm"), y = unit(1, "npc") - unit(1, "lines"), just = c("left",
                                                                                                                     "bottom"), vp = vpPath("parent", "plot"), name = "rawKeyText",
                                 gp = gpar(cex = 0.9, fontface = "bold"))
    stringw <- stringWidth(grobs$rawKeyText$label)
    grobs$fitKey = linesGrob(x = unit(c(19, 25), "mm") + stringw,
                             y = unit(1, "npc") - unit(0.6, "lines"), vp = vpPath("parent",
                                                                                  "plot"), name = "trendKey", gp = gpar(col = fit.col,
                                                                                                                        lwd = 3))
    grobs$fitKeyText <- textGrob(label = "Fitted", x = unit(27,
                                                            "mm") + stringw, y = unit(1, "npc") - unit(1, "lines"),
                                 just = c("left", "bottom"), vp = vpPath("parent", "plot"),
                                 name = "trendKeyText", gp = gpar(cex = 0.9, fontface = "bold"))
    stringw <- stringw + stringWidth(grobs$fitKeyText$label)
    grobs$predKey <- linesGrob(x = unit(c(36, 42), "mm") + stringw,
                               y = unit(1, "npc") - unit(0.6, "lines"), vp = vpPath("parent",
                                                                                    "plot"), name = "trendSeasonKey", gp = gpar(col = pred.col,
                                                                                                                                lwd = 3))
    grobs$predKeyText <- textGrob(label = "Prediction", x = unit(44,
                                                                 "mm") + stringw, y = unit(1, "npc") - unit(1, "lines"),
                                  just = c("left", "bottom"), vp = vpPath("parent", "plot"),
                                  name = "trendSeasonKeyText", gp = gpar(cex = 0.9, fontface = "bold"))
    tree <- gTree(children = grobs, childrenvp = vp.tree, name = "tree")


    if (show){
      newdevice(width = 7, height = 4)
      drawImage(tree)
    }
    pred
    }
